test/pstl_offload/algorithms/algorithms_offload.pass.cpp

```diff
@@ -0,0 +1,1611 @@
+{
+    store_id(algorithm_id::TRANSFORM_INCLUSIVE_SCAN_INIT);
+    check_policy(__exec);
+    return __result;
+}
+
+template <typename _ExecutionPolicy, typename _ForwardIterator2, typename _UnaryOperation, typename _BinaryOperation>
+test_enable_if_execution_policy<_ExecutionPolicy, _ForwardIterator2>
+transform_inclusive_scan(_ExecutionPolicy&& __exec, not_iterator __first, not_iterator __last,
+                         _ForwardIterator2 __result, _BinaryOperation __binary_op, _UnaryOperation __unary_op)
+{
+    store_id(algorithm_id::TRANSFORM_INCLUSIVE_SCAN);
+    check_policy(__exec);
+    return __result;
+}
+
+template <typename _ExecutionPolicy, typename _ForwardIterator2, typename _BinaryOperation>
+test_enable_if_execution_policy<_ExecutionPolicy, _ForwardIterator2>
+adjacent_difference(_ExecutionPolicy&& __exec, not_iterator __first, not_iterator __last,
+                    _ForwardIterator2 __d_first, _BinaryOperation __op)
+{
+    store_id(algorithm_id::ADJACENT_DIFFERENCE_BINARYOP);
+    check_policy(__exec);
+    return __d_first;
+}
+
+template <typename _ExecutionPolicy, typename _ForwardIterator2>
+test_enable_if_execution_policy<_ExecutionPolicy, _ForwardIterator2>
+adjacent_difference(_ExecutionPolicy&& __exec, not_iterator __first, not_iterator __last,
+                    _ForwardIterator2 __d_first)
+{
+    store_id(algorithm_id::ADJACENT_DIFFERENCE);
+    check_policy(__exec);
+    return __d_first;
+}
+
+} // namespace oneapi::dpl
+
+#include <oneapi/dpl/execution>
+
+#include <algorithm>
+#include <memory>
+#include <numeric>
+#include <execution>
+
+#include <support/utils.h>
+#include <support/utils_invoke.h>
+
+void store_id(algorithm_id __id) {
+    EXPECT_TRUE(algorithm_id_state == algorithm_id::EMPTY_ID, "algorithm_id contains non empty value");
+    algorithm_id_state = __id;
+}
+
+template <typename _T>
+struct is_device_policy : std::false_type {};
+
+template <typename _KernelName>
+struct is_device_policy<oneapi::dpl::execution::device_policy<_KernelName>> : std::true_type {};
+
+template <typename _ExecutionPolicy>
+void check_policy(const _ExecutionPolicy& __policy) {
+    static_assert(is_device_policy<std::decay_t<_ExecutionPolicy>>::value, "Algorithm was not offloaded to the device");
+    EXPECT_TRUE(__policy.queue().get_device() == TestUtils::get_pstl_offload_device(), "Algorithm was offloaded to the wrong device");
+}
+
+template <algorithm_id _AlgorithmId, typename _RunAlgorithmBody, typename... _AlgorithmArgs>
+void test_algorithm(_RunAlgorithmBody __run_algorithm, _AlgorithmArgs&&... __args) {
+    EXPECT_TRUE(algorithm_id_state == algorithm_id::EMPTY_ID, "algorithm_id was not reset");
+    __run_algorithm(std::execution::par_unseq, std::forward<_AlgorithmArgs>(__args)...);
+
+    EXPECT_TRUE(algorithm_id_state == _AlgorithmId, "Algorithm was not offloaded to the device or incorrect oneDPL algorithm was called");
```

Comment:
Because we handle two possible types of errors. For example

- If `std::sort(par_unseq)` didn't call `oneapi::dpl::sort` and standard implementation was used- the algorithm was not offloaded
- If `std::sort(par_unseq)` called `oneapi::dpl::sort` with predicate or by mistake `oneapi::dpl::for_each`  - it means that incorrect oneDPL algorithm was called.

Reply:
Replaced with two separate assertions